File: coin_change/coin_change.py
# uv run coin_change.py

# /// script
# requires-python = ">=3.13"
# dependencies = []
# ///


def coinChange(coins, amount):
    memo = {}
    inf = float('inf')

    def solve(remaining):
        if remaining == 0: return 0
        if remaining < 0: return -1
        if remaining in memo: return memo[remaining]

        min_coins = inf

        for coin in coins:
            res = solve(remaining - coin)
            if res != -1:
                min_coins = min(min_coins, 1 + res)

        memo[remaining] = min_coins if min_coins != inf else -1
        return memo[remaining]

    return solve(amount)

def test():
    tests = [
        [
            [
                [1,2,5],
                11
            ],
            3
        ],
        [
            [
                [2],
                3
            ],
            -1
        ],
        [
            [
                [1],
                0
            ],
            0
        ],
        [
            [
                [186,419,83,408],
                6249
            ],
            20
        ],
    ]

    for input, want in tests:
        res = coinChange(*input)
        if want != res:
            raise RuntimeError(f"Unexpected response, want={want}, got={res}")

if __name__ == "__main__":
    test()

File: coin_change/coin_change.ts
// node coin_change.ts

type Memo = { [key: string]: number };

function coinChange(coins: number[], amount: number): number {
  const memo: Memo = {};

  function solve(remaining: number): number {
    if (remaining == 0) return 0;
    if (remaining < 0) return -1;
    if (memo[remaining]) return memo[remaining];

    let min_count = Infinity;

    for (const coin of coins) {
      const res = solve(remaining - coin);
      if (res != -1) {
        min_count = Math.min(min_count, 1 + res);
      }
    }

    memo[remaining] = min_count == Infinity ? -1 : min_count;
    return memo[remaining];
  }

  return solve(amount)
}

type Input = Parameters<typeof coinChange>;
type Test = [Input, number];

function test() {
  const tests: Test[] = [
    [
        [
            [1,2,5],
            11
        ],
        3
    ],
    [
        [
            [2],
            3
        ],
        -1
    ],
    [
        [
            [1],
            0
        ],
        0
    ],
    [
        [
            [186,419,83,408],
            6249
        ],
        20
    ],
  ];

  for (const [input, exp] of tests) {
    const res = coinChange(...input);
    if (exp != res) {
      throw new Error(`Unexpected result, want=${exp}, got=${res}`);
    }
  }
}

test()